### ISSUE TITLE
Unknown warning option -Wno-maybe-uninitialized and c++ compile errors with blis.h

### ISSUE THREAD

**Issue by fxkaiser** [flame/blis#309]
Hey guys,

i wanted to try out blis as backend for my own c++ tensor template library. Configure yields

>*configure: using 'clang' compiler.
configure: found clang version 10.0.1 (maj: 10, min: 0, rev: 1).
configure: checking for blacklisted configurations due to clang 10.0.1.
configure: found assembler ('as') version 10.0.1 (maj: 10, min: 0, rev: 1).
configure: checking for blacklisted configurations due to as 10.0.1.
configure: reading configuration registry...done.
configure: checking whether we need to update the version file.
configure: checking version file './version'.
configure: found '.git' directory; assuming git clone.
configure: executing: git describe --tags.
configure: got back 0.3.1-21-ga731a428.
configure: truncating to 0.3.1-21.
configure: updating version file './version'.
configure: starting configuration of BLIS 0.3.1-21.
configure: configuring with official version string.
configure: automatic configuration requested.
configure: hardware detection driver returned 'haswell'.
configure: checking configuration against contents of 'config_registry'.
configure: configuration 'haswell' is registered.
configure: 'haswell' is defined as having the following sub-configurations:
configure:    haswell
configure: which collectively require the following kernels:
configure:    haswell zen
configure: checking sub-configurations:
configure:   'haswell' is registered...and exists.
configure: checking sub-configurations' requisite kernels:
configure:   'haswell' kernels...exist.
configure:   'zen' kernels...exist.
configure: using install prefix '/usr/local/'.
configure: debug symbols disabled.
configure: disabling verbose make output. (enable with 'make V=1'.)
configure: building BLIS as a static library.
configure: threading is disabled.
configure: internal memory pools for packing buffers are enabled.
configure: libmemkind not found; disabling.
configure: the BLAS compatibility layer is enabled.
configure: the CBLAS compatibility layer is enabled.
configure: the internal integer size is automatically determined.
configure: the BLAS/CBLAS interface integer size is 32-bit.
...*

and running make check 

>*Compiling obj/haswell/blastest/cblat1.o
warning: unknown warning option '-Wno-maybe-uninitialized'; did you mean '-Wno-uninitialized'? [-Wunknown-warning-option]
1 warning generated.
Compiling obj/haswell/blastest/abs.o
warning: unknown warning option '-Wno-maybe-uninitialized'; did you mean '-Wno-uninitialized'? [-Wunknown-warning-option]
1 warning generated.
Compiling obj/haswell/blastest/acos.o
warning: unknown warning option '-Wno-maybe-uninitialized'; did you mean '-Wno-uninitialized'? [-Wunknown-warning-option]
1 warning generated.
...*

It would be nice if these warnings could be eliminated by using the correct warning option.

Since I already post this issue here I would like to ask a question (I know that the google mailing list would be the right place for this, sorry).

I wrote a simple c++ test program (compiling using clang++, version as above) including

a) blis.h (only) to have direct access to the dgemm micro kernel, but compilation fails with the following errors:

>*clang++ blis_test.cpp -o blis_test -I /usr/local/include/ -lblis -lpthread
In file included from blis_test.cpp:4:
/usr/local/include/blis/blis.h:1135:9: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
typedef struct mtx_s
        ^
/usr/local/include/blis/blis.h:17630:15: error: cannot initialize a variable of type 'const num_t' with an rvalue of type 'unsigned int'
        const num_t  dt    = bli_obj_datatype( *obj );
                     ^       ~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:17675:11: error: assigning to 'num_t' from incompatible type 'int'
            dt = bli_datatype_proj_to_real( dt );
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:4848:2: note: expanded from macro 'bli_datatype_proj_to_real'
        ( dt & ~BLIS_BITVAL_COMPLEX )
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:17685:11: error: assigning to 'num_t' from incompatible type 'int'
            dt = bli_datatype_proj_to_real( dt );
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:4848:2: note: expanded from macro 'bli_datatype_proj_to_real'
        ( dt & ~BLIS_BITVAL_COMPLEX )
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:17692:15: error: cannot initialize a variable of type 'const num_t' with an rvalue of type 'unsigned int'
        const num_t  dt    = bli_obj_datatype( *obj );
                     ^       ~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:17789:10: error: cannot initialize a variable of type 'func_t *' (aka 'func_s *') with an rvalue of type 'void *'
        func_t* func = bli_cntx_get_packm_ker_dt( dt, ker_id, cntx );
                ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:17803:10: error: cannot initialize a variable of type 'func_t *' (aka 'func_s *') with an rvalue of type 'void *'
        func_t* func = bli_cntx_get_unpackm_ker_dt( dt, ker_id, cntx );
                ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:18363:9: error: cannot initialize return object of type 'void *' with an lvalue of type 'malloc_ft' (aka 'void *(*)(unsigned long)')
        return membrk->malloc_fp;
               ^~~~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:18368:9: error: cannot initialize return object of type 'void *' with an lvalue of type 'free_ft' (aka 'void (*)(void *)')
        return membrk->free_fp;
               ^~~~~~~~~~~~~~~
/usr/local/include/blis/blis.h:18375:22: error: assigning to 'malloc_ft' (aka 'void *(*)(unsigned long)') from incompatible type 'void *'
        membrk->malloc_fp = malloc_fp;
                            ^~~~~~~~~
/usr/local/include/blis/blis.h:18380:20: error: assigning to 'free_ft' (aka 'void (*)(void *)') from incompatible type 'void *'
        membrk->free_fp = free_fp;
                          ^~~~~~~
1 warning and 10 errors generated.*

b) cblas.h (only) to call cblas_dgemm as a test, this works fine

Is this intendend ? Am i missing something ? My solution would be to build my own backend that links against libblis but has an c++ friendly api.

Best regards,
fxkaiser

**Comment by fgvanzee** [comment]
@fxkaiser Thanks for your report, and sorry to hear about your compilation issues. I'll let @devinamatthews comment on your C++ issues.

As for the warnings related to `-Wno-maybe-uninitialized`: The real irony here is that that option is meant to silence warnings that are output by the compiler when building the BLAS test drivers (ie: which were not authored by us). However, as far as I can tell, that option is not added unless `CC_VENDOR` is `gcc`. That logic may not be available in your case, though, because you are using a version of BLIS that is so old. May I ask why you are using a version based on 0.3.1?

**Comment by devinamatthews** [comment]
@fxkaiser it's been a while since I tried to use blis.h with a C++ compiler and it seems the list of incompatibilities has grown. I will try to take a look at this in the near future. I imagine many or possibly all of these can be silenced by a sufficiently permissible set of compiler options if you want to try a quick and dirty test.

**Comment by fxkaiser** [comment]
@devinamatthews @fgvanzee There is no reason I used 0.3.1. Actually I want to use the latest release. Thought I updated it but I didn't. Now, I cloned the latest version 0.5.2 and repeated the configure, make, make check, make install steps once again. This time everything works fine. The warnings mentioned above are gone (even with clang), and blis.h makes no problem if used with clang++. Sorry for any inconveniences caused by my naive post, should have checked that in advance.

Best regards,
fxkaiser
